pkg/blueprint/partitioning_customizations.go

```diff
@@ -0,0 +1,545 @@
+	decoder.DisallowUnknownFields()
+	if err := decoder.Decode(&vg); err != nil {
+		return fmt.Errorf("error decoding partition with type \"lvm\": %w", err)
+	}
+
+	v.VGCustomization = vg.VGCustomization
+	return nil
+}
+
+// Custom TOML unmarshaller that first reads the value of the "type" field and
+// then deserialises the whole object into a struct that only contains the
+// fields valid for that partition type. This ensures that no fields are set
+// for the substructure of a different type than the one defined in the "type"
+// fields.
+func (v *PartitionCustomization) UnmarshalTOML(data any) error {
+	errPrefix := "TOML unmarshal:"
+	d, ok := data.(map[string]any)
+	if !ok {
+		return fmt.Errorf("%s customizations.partition is not an object", errPrefix)
+	}
+
+	partType := "plain"
+	if typeField, ok := d["type"]; ok {
+		typeStr, ok := typeField.(string)
+		if !ok {
+			return fmt.Errorf("%s type must be a string, got \"%v\" of type %T", errPrefix, typeField, typeField)
+		}
+		partType = typeStr
+	}
+
+	// serialise the data to JSON and reuse the subobject decoders
+	dataJSON, err := json.Marshal(data)
+	if err != nil {
+		return fmt.Errorf("%s error while decoding partition customization: %w", errPrefix, err)
+	}
+	switch partType {
+	case "plain":
+		if err := decodePlain(v, dataJSON); err != nil {
+			return fmt.Errorf("%s %w", errPrefix, err)
+		}
+	case "btrfs":
+		if err := decodeBtrfs(v, dataJSON); err != nil {
+			return fmt.Errorf("%s %w", errPrefix, err)
+		}
+	case "lvm":
+		if err := decodeLVM(v, dataJSON); err != nil {
+			return fmt.Errorf("%s %w", errPrefix, err)
+		}
+	default:
+		return fmt.Errorf("%s unknown partition type: %s", errPrefix, partType)
+	}
+
+	v.Type = partType
+
+	if minsizeField, ok := d["minsize"]; ok {
+		minsize, err := decodeSize(minsizeField)
+		if err != nil {
+			return fmt.Errorf("%s error decoding minsize for partition: %w", errPrefix, err)
+		}
+		v.MinSize = minsize
+	}
+
+	return nil
+}
+
+// Validate checks for customization combinations that are generally not
+// supported or can create conflicts, regardless of specific distro or image
+// type policies. The validator ensures all of the following properties:
+//   - All mountpoints are valid
+//   - All mountpoints are unique
+//   - All LVM volume group names are unique
+//   - All LVM logical volume names are unique within a given volume group
+//   - All btrfs subvolume names are unique within a given btrfs volume
+//   - All btrfs subvolume names are valid and non-empty
+//   - All filesystems are valid for their mountpoints (e.g. xfs or ext4 for /boot)
+//   - No LVM logical volume has an invalid mountpoint (/boot or /boot/efi)
+//   - Plain filesystem types are valid for the partition type
+//   - All non-empty properties are valid for the partition type (e.g.
+//     LogicalVolumes is empty when the type is "plain" or "btrfs")
+func (p *DiskCustomization) Validate() error {
+	if p == nil {
+		return nil
+	}
+
```

Comment:
We discussed this a bit in the previous PR: https://github.com/osbuild/images/pull/926#discussion_r1820296291

Copying from https://github.com/osbuild/images/pull/926#discussion_r1822786092:
> The problem I was trying to solve is being able to validate without the layout constraints (but also being able to validate with the constraints, either separately or at the same time), so that the blueprint code itself applies the extra constraints (at the UI level) but the internals (NewCustomPartitionTable()) only check for invalid configurations like conflicting names and labels.

I want to be able to validate the customizations with `Validate()` in the `NewCustomPartitionTable()` function for just the things that are actually invalid.  So things like multiple LVM VGs should be allowed when calling `NewCustomPartitionTable()` directly.  But the "layout constraints" considered by `ValidateLayoutConstraints()` should check for the things we want to disallow as a matter of support policy in osbuild-composer and bootc-image-builder.

Reply:
Added some comments in https://github.com/osbuild/images/pull/1064